WIP: Add support for automatic discovery of TP-Link switches and bulbs

rytilahti opened the pull request:
## Description:
This PR adds support for automatic discovery of TP-Link bulbs and switches, based on the new config flow. The integration needs to be enable either in the integration configurations or alternatively by adding `tplink` component to the configuration file.

Breaking change: the device name is not configurable over configuration file but is read from the device. To change the name it has to be either changed inside homeassistant, or alternatively on the device itself.

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

rytilahti committed:
{switch,light}.tplink: use deviceid as unique id, fetch name from the device during initialization

rytilahti committed:
raise PlatformNotReady when no device is available

rytilahti commented:
@DanNixon ping, this changes the name fetching, does it look okay to you?

DanNixon commented:
:+1: Looks good to me.

MartinHjelmare reviewed (approved):
Looks good!

This looks like a breaking change due to the name change. Please add a paragraph to the PR description about this for the release notes, or tell me why I'm wrong.

amelchio commented:
Is it possible/advisable to use the device MAC address? That is printed on my plug and would allow for easy identification.

rytilahti committed:
Use mac instead of deviceid

rytilahti committed:
remove name option as obsolete

rytilahti committed:
Add support for configuration flow / integration

Allows activating automatic discovery of supported devices from the configuration

rytilahti committed:
Fix linting, update requirements_all.txt

awarecan commented on homeassistant/components/tplink/__init__.py:
```diff
@@ -0,0 +1,62 @@
+
+
+async def async_setup_entry(hass, entry):
+    """Set up Sonos from a config entry."""
```
Typo in comment

awarecan commented on homeassistant/components/tplink/__init__.py:
```diff
@@ -0,0 +1,62 @@
+    """Set up Sonos from a config entry."""
+    from pyHS100 import Discover, SmartBulb, SmartPlug
+
+    hass.data[DOMAIN] = {'light': [], 'switch': []}
```
What if user has manual configured devices already imported, will this overwrite it?

rytilahti replied:
The previous mechanism does not use any part of this, so it has no idea about auto-detected devices, but the ones which came after would error out due to a clash with unique ids.

What would be the best way to handle this? I'm unsure if this is going to be initialized always and what's the initialization order. And if this is not initialized always, where should the dictionary be initialized? I could adjust the `setup_platform()`s to check for an existing device and bail out if one already exists, but the initialization should happen somewhere else. 

This all assumes that the initialization order is 1) manually configured devices from configuration.yaml etc. and first after that 2) auto-detected devices. Is this a correct assumption?

awarecan replied:
I think we should respect user configured device, ignore duplicated finding in discovery. Our users are so smart, I had seen complaints about cannot discovery Sonos or Cast (since they put them in separate VLAN etc), and they are asking keep the capability to add device through configuraiton.yaml.

rytilahti replied:
Yes, an ability to configure these locally is a must have of course. The information stored in `hass.data` is used only for auto-detected devices at the moment (it stores already initialized instances), so I think the proper way to fix this is:
1. Add manually configured devices to those lists.
2. Check the list for unique id to before initializing the discovered devices.

The remaining question is where should that dictionary be initialized?

balloob replied:
Things should be initialized as close to where it is going to be used. So the `switch.tplink` platform will only use it, it should live there.

balloob replied:
Oh I see now that we do discovery once in the component, in that case initialize it here.

balloob commented on homeassistant/components/light/tplink.py:
```diff
@@ -18,6 +18,9 @@
     color_temperature_kelvin_to_mired as kelvin_to_mired)
+from homeassistant.exceptions import PlatformNotReady
+
+DOMAIN = 'tplink'
```
This is not good. You need to import `DOMAIN` and rename it:

```python
from homeassistant.components.tplink import DOMAIN as TPLINK_DOMAIN
```

balloob commented on homeassistant/components/light/tplink.py:
```diff
@@ -18,6 +18,9 @@
+
+DOMAIN = 'tplink'
 
 REQUIREMENTS = ['pyHS100==0.3.2']
```
This is not needed. This is installed by the component.

balloob replied:
Also add `DEPENDENCIES = ['tplink']`

balloob commented on homeassistant/components/tplink/__init__.py:
```diff
@@ -0,0 +1,62 @@
+
+    if hass.data[DOMAIN]['light']:
+        hass.async_add_job(
+            hass.config_entries.async_forward_entry_setup(entry, 'light'))
```
You did not add an `async_setup_entry` method to `light.tplink`.

balloob commented on homeassistant/components/tplink/__init__.py:
```diff
@@ -0,0 +1,62 @@
+    return await hass.async_add_executor_job(discover)
+
+
+config_entry_flow.register_discovery_flow(DOMAIN,
```
Adding a config flow requires tests to be added.

balloob commented on homeassistant/components/tplink/__init__.py:
```diff
@@ -0,0 +1,62 @@
+
+    hass.data[DOMAIN] = {'light': [], 'switch': []}
+    devices = await hass.async_add_executor_job(Discover.discover)
+    for dev in devices.values():
```
Why not use `_async_has_devices`?

balloob commented on homeassistant/components/tplink/__init__.py:
```diff
@@ -0,0 +1,62 @@
+
+async def async_setup(hass, config):
+    """Set up the TP-Link component."""
+    conf = config.get(DOMAIN)
```
Why are you doing this? When you copy paste code, you need to understand what it does.

Sonos and Cast have this line so that people can still configure the platforms how they used to

rytilahti commented:
Thanks a lot for the review @balloob, the points you make are gvery helpful. This PR should have been marked as a WIP in the first place, as it was not clear to me whether to break the existing configuration syntax in favor of "component-based" one as you suggested. I'll dig into making the changes after I return from holidays.

balloob commented:
This PR seems to have gone stale. Closing it.

rytilahti commented:
I have revised the code based on the previous feedback so I'm reopening this.

> * The code should be structured such that a config entry _always_ exists.

I think this is now handled.

> * Users can configure extra lights and platforms under the `tplink` component.
>   ```yaml
>   tplink:
>     light:
>     - host: 192.168.0.100
>     - host: 192.168.0.120
>     switch:
>     - host: 192.168.0.130
>     - host: 192.168.0.140
>   ```

Done. The configuration file overrides the detection (which could have been useful for dimmers, which are detected as switches, but act as lights).

> * During initialization of `tplink` component, create an entry via import and store the `tplink` config in `hass.data`

Done.

> * During component config entry setup, check if `light` or `switch` are configured, if not, run discovery. Check which platforms are needed and forward entry.

I think it's better to run discovery even when there are configured lights (think of having devices in undiscoverable networks, which need to be manually configured).

> * `async_setup_platform` should be removed for both tplink light and switch. Instead have `async_setup_entry` that will now configure all devices at once. `async_setup_entry` is async, so don't execute sync code. Instead fetch it inside `update` method.

Done, there is one problem though. `async_setup_entry` uses `PlatformNotReady` in case it fails to query the device. As the devices are now passed all at once, having a single non-responsive device will cause abort on all devices for that class. Should a logger be used here instead of raising that exception? 

> * Add `PARALLEL_UPDATES = 0` to both `light` and `switch` platform so that updates are run in parallel (since each entity hits a unique device)

Done.
